### ISSUE TITLE
Improved FFI integration

### ISSUE THREAD

**Issue by UlrichEckhardt** [vimeo/psalm#9230]
Heya!

_Note up front: I'm not too familiar with either of Psalm or FFI, just trying to solve an issue there._

The problem I have is for code like this:
```
$ffi = FFI::cdef(file_get_contents($configuration->getHeaderFilePath()), $configuration->getLibraryFilePath());
$instance = $ffi->createInstance();
```
Psalm complains about an `UndefinedMethod` for the `$ffi->createInstance()` call, which is understandable, because the call is evaluated dynamically. What I did in order to "fix" that was to modify vendor/vimeo/psalm/stubs/extensions/ffi.phpstub:
```
namespace {
    final class FFI
    {
        ... // unchanged to dist

        // new method stub
        public function __call(...$values)
        {
        }
    }
}
```
Any comments on that approach?

Just for completeness, FFI also needs `__get()` and `__set()` (and perhaps `__isset()`?) [magic methods](https://www.php.net/manual/en/language.oop5.overloading.php). I don't have a use case for that, so it's secondary to me. Still, I created some code, added those to the ffi.phpstub, but couldn't stop Psalm from complaining. Perhaps I'll look further into it tomorrow (or today by now, if you catch my drift).

Cheers and good night!

Uli

**Comment by orklah** [comment]
That seems to make sense, you could try to create a PR and some tests and we'll see how Psalm behaves and how we can fix potential errors

**Comment by UlrichEckhardt** [comment]
Yep, easier than I thought: https://psalm.dev/r/ffb1bc3821. The code was taken straight from the examples on https://www.php.net/manual/en/ffi.examples-basic.php.

**Comment by weirdan** [comment]
Seems to be fixed (in #9368 ?)
